### ISSUE TITLE
Support for Aquastream Ultimate

### ISSUE THREAD

**Issue by Alexander-LIT** [aleksamagicka/aquacomputer_d5next-hwmon#50]
I have both an Aquero 6XT as well as an Aquastream Ultimate connected via USB.

The data from the Aquero is listed when i run 'sensors' and works great, however the Pump is not showing up. I figured since Aquastream XT is supported, perhaps the Ultimate would work as well. 
If there is any more information that I can supply to help solve this issue please let me know.

make dev                                                                  
make clean
make[1]: Entering directory '/home/alex/aquacomputer_d5next-hwmon'
make -C /lib/modules/6.0.8-1-MANJARO/build M=$PWD clean
make[2]: Entering directory '/usr/lib/modules/6.0.8-1-MANJARO/build'
  CLEAN   /home/alex/aquacomputer_d5next-hwmon/Module.symvers
make[2]: Leaving directory '/usr/lib/modules/6.0.8-1-MANJARO/build'
make[1]: Leaving directory '/home/alex/aquacomputer_d5next-hwmon'
make
make[1]: Entering directory '/home/alex/aquacomputer_d5next-hwmon'
make -C /lib/modules/6.0.8-1-MANJARO/build M=$PWD modules
make[2]: Entering directory '/usr/lib/modules/6.0.8-1-MANJARO/build'
  CC [M]  /home/alex/aquacomputer_d5next-hwmon/aquacomputer_d5next.o
  MODPOST /home/alex/aquacomputer_d5next-hwmon/Module.symvers
  CC [M]  /home/alex/aquacomputer_d5next-hwmon/aquacomputer_d5next.mod.o
  LD [M]  /home/alex/aquacomputer_d5next-hwmon/aquacomputer_d5next.ko
  BTF [M] /home/alex/aquacomputer_d5next-hwmon/aquacomputer_d5next.ko
make[2]: Leaving directory '/usr/lib/modules/6.0.8-1-MANJARO/build'
make[1]: Leaving directory '/home/alex/aquacomputer_d5next-hwmon'
sudo rmmod aquacomputer_d5next || true
[sudo] password for alex: 
sudo insmod aquacomputer_d5next.ko

lsusb
Bus 012 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 011 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 009 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 010 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 008 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 007 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 006 Device 002: ID 174c:3074 ASMedia Technology Inc. ASM1074 SuperSpeed hub
Bus 006 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 005 Device 005: ID 0c70:f00b MCT Elektronikladen aquastream ULTIMATE
Bus 005 Device 004: ID 0c70:f001 MCT Elektronikladen aquaero
Bus 005 Device 003: ID 174c:2074 ASMedia Technology Inc. ASM1074 High-Speed hub
Bus 005 Device 002: ID 046d:c539 Logitech, Inc. Cordless Mouse Receiver
Bus 005 Device 006: ID 0b05:18f3 ASUSTek Computer, Inc. AURA LED Controller
Bus 005 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 004 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 003 Device 004: ID 30be:1014 Schiit Audio Schiit Modi 3+
Bus 003 Device 003: ID 0603:0002 Novatek Microelectronics Corp. Sino Wealth keyboard/mouse 2.4 GHz receiver
Bus 003 Device 002: ID 3297:4976 ZSA Technology Labs Inc ErgoDox EZ Glow
Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub

**Comment by aleksamagicka** [comment]
From your output: `Bus 005 Device 005: ID 0c70:f00b MCT Elektronikladen aquastream ULTIMATE`, so the product ID is `0xf00b`. The Aquastream XT that the driver currently supports has the ID `0xf0b6`.

Looks like it was released in ~2015/2016, so it should use the regular HID report communication (old Aquacomputer stuff, like the aquastream xt, communicates through a legacy way). I've looked it up on Youtube, the interface on the OLED looks extremely similar to the one on the D5 Next, so they probably share a similar base.

Unfortunately without you looking up the HID reports that the pump sends and pairing that with sensor data from aquasuite/hwinfo64 (so the reported values match up in the reports), I can't do anything reliably from my side. If you want to delve in I can send you a hints/tutorial doc for sniffing USB traffic.

**Comment by Alexander-LIT** [comment]
I have never done anything like that, but certainly willing to give it a try.

**Comment by aleksamagicka** [comment]
[Here](https://github.com/liquidctl/liquidctl/blob/main/docs/developer/capturing-usb-traffic.md) it is. For you, the filter should be `usb.idVendor == 0x0c70 && usb.idProduct == 0xf00b`.

If it's similar to the D5 Next, it should send a status HID report every second. If you need assistance, write here and I'll try my best to help.

**Comment by Alexander-LIT** [comment]
Okay so here goes:

The tutorial is slightly confusing as they are using the Linux host to capture the packets, which in my case did not really work (or I was doing it wrong).

I installed wireshark on my Linux host, and ran mobprobe usbmon to get the usb devices to show up in wireshark. Selected all of them, but when filtering nothing showed up (with the VM running). I assume that is because the device is passed through to the windows VM and thus not available on the host. 

As a next step I installed Wireshark inside the Win10 VM and was able to filter the packets and save them. What would be the next steps? You mentioned pairing the sensor data, or is it sufficient if I send you the captures?

Cheers!

**Comment by aleksamagicka** [comment]
If a device is passed in to a VM, it won't be visible on the host.

You can send the reports here, of course, but to do anything we need the readings from hwinfo64/aquasuite at the time of the report. What I mean by this is that the report contains the readings of the sensors (temp is 25C, fan speed is 800RPM etc.), but to find where those values are located _in_ the report, we need to know their values - aquasuite already knows how to parse the report, and the data it outputs is what we need to reverse engineer their positions in the hex dump (pairing the sensor data).

Long story short 😄 - to do anything useful, send the reports you captured and a screenshot of aquasuite/hwinfo64 readings so I can try and look them up.

**Comment by Alexander-LIT** [comment]
Ok so we are looking for a fairly short capture with a screenshot of the data from the time of capture. On it!

**Comment by aleksamagicka** [comment]
That's right. Here's [how it looks](https://github.com/liquidctl/liquidctl/blob/main/docs/developer/protocol/aquacomputer.md#d5-next-pump) on the D5 Next.

**Comment by Alexander-LIT** [comment]
Here is to hoping I did it correctly :-)

[aquastream-ultimate1.zip](https://github.com/aleksamagicka/aquacomputer_d5next-hwmon/files/10124763/aquastream-ultimate1.zip)

**Comment by aleksamagicka** [comment]
Thanks. Here goes nothing...

| What                                    | Where | Value              |
| --------------------------------------- | ----- | ------------------ |
| Serial key, first part                  | 0x03  | 16293              |
| Serial key, second part                 | 0x05  | 20503              |
| Firmware version                        | 0xD   | 1011               |
| Water temperature                       | 0x2D  | 2999 (29.99C)      |
| *First virtual (software) temp sensor*  | 0x2F  | Empty/unset        |
| *Fourth virtual (software) temp sensor* | 0x35  | Set to water temp? |
| *Fifth virtual (software) temp sensor*  | 0x37  | Set to 0?          |
| Pump +12V                               | 0x3D  | 1222 (12.22V)      |
| Fan +12V                                | 0x43  | 1127 (11.27V)      |
| Pump RPM                                | 0x51  | 4790 (RPM)         |
| Pump current                            | 0x53  | 673 (0.673A)       |
| Pump power                              | 0x55  | 822 (8.22W)        |
| Pressure                                | 0x57  | 305 (0.305bar)     |

Can you please confirm the info above (serial number, firmware)? How many virtual sensors does it have?

**Comment by aleksamagicka** [comment]
Thanks. In that case I won't touch the supposed virtual sensors, and will implement the rest. I will write here when I implement that in a separate branch for you to test.

**Comment by Alexander-LIT** [comment]
[aquero.zip](https://github.com/aleksamagicka/aquacomputer_d5next-hwmon/files/10125380/aquero.zip)

I tried a capture of the aquero (with f001 filter) and took another screenshot.

**Comment by aleksamagicka** [comment]
Thanks, I found it in the sensor report (`0xF9`). Will make that adjustment as well.

**Comment by Alexander-LIT** [comment]
Wonderful! Thank you so much!

**Comment by aleksamagicka** [comment]
Here's the [aquaero test branch](https://github.com/aleksamagicka/aquacomputer_d5next-hwmon/tree/aquaero-flow-sensor-1), for the flow sensor.

**Comment by Alexander-LIT** [comment]
Sorry about my beginner questions... how do I clone a different branch?

In the meantime, I just copied the changed file:

aquaero-hid-3-9
Adapter: HID adapter
Fan 1 voltage:         12.07 V  
Fan 2 voltage:         12.26 V  
Fan 3 voltage:         12.15 V  
Fan 4 voltage:          0.00 V  
Fan 5 voltage:          0.00 V  
Fan 1 speed:          1000 RPM  (min =  300 RPM, max = 1800 RPM)
Fan 2 speed:           556 RPM  (min =  300 RPM, max = 1800 RPM)
Fan 3 speed:           749 RPM  (min =  500 RPM, max = 1500 RPM)
Fan 4 speed:             0 RPM  (min =  450 RPM, max = 2000 RPM)
**Flow sensor 1 [dl/h]: 1281 RPM**
Sensor 1:              +30.4°C  
Sensor 2:              +25.5°C  
Sensor 3:              +43.5°C  
Sensor 4:                  N/A  
Sensor 5:                  N/A  
Sensor 6:                  N/A  
Sensor 7:                  N/A  
Sensor 8:                  N/A  
Virtual sensor 1:      +50.0°C  
Virtual sensor 2:      +50.0°C  
Virtual sensor 3:      +50.0°C  
Virtual sensor 4:      +50.0°C  
Virtual sensor 5:      +50.0°C  
Virtual sensor 6:      +50.0°C  
Virtual sensor 7:      +50.0°C  
Virtual sensor 8:      +50.0°C  
Fan 1 power:            2.58 W  
Fan 2 power:            0.00 W  
Fan 3 power:            1.40 W  
Fan 4 power:            0.00 W  
Fan 5 power:            0.00 W  
Fan 1 current:        214.00 mA 
Fan 2 current:          0.00 A  
Fan 3 current:        116.00 mA 
Fan 4 current:          0.00 A  
Fan 5 current:          0.00 A

**Comment by aleksamagicka** [comment]
You can use `git checkout`:

```
git checkout -b my_new_local_branch origin/the-remote-branch-i-want
```

In this case, something like this, depending on how you want to name your local branch:

```
git checkout -b aquaero-flow-sensor-1 origin/aquaero-flow-sensor-1
```

**Comment by aleksamagicka** [comment]
And [here](https://github.com/aleksamagicka/aquacomputer_d5next-hwmon/tree/aquastream-ultimate-support) is the new branch for Aquastream Ultimate. Please test and report if anything is missing or badly reported.

**Comment by Alexander-LIT** [comment]
Seems to work like a charm.

I compared the data from "watch sensors" with whats on the Aqueros' display and other than different update timings between my display and the terminal, the values seem spot on.

Maybe "RPM" could be removed from the output and values converted to liters/h?

**Comment by aleksamagicka** [comment]
Great, thank you!

Others have had the same idea before, but "RPM" is kind of the catch all type in hwmon (internally, it's regarded as fan output), and it's the only of the bunch (temp, RPM, voltage, current, humidity, etc.) that I can put general values in (such as flow and pressure) since it doesn't carry a too exact meaning, like temp would for example. And it's also accepted that way into the kernel. Also, "RPM" is shown by the `sensors` command, not by the driver, since it's of type fan; the driver provides only the label and value alongside that.

For liters/h, aside from the above, we run into an another problem: first, sysfs only exposes whole numbers, and second, generally no floating point operations are possible in the kernel, except if you are on x86 and put your code between `kernel_fpu_start()` and `kernel_fpu_end()` (which is to say, in very special occassions). So, to not lose precision given by the device, we expose it as dL/h. Same thing with the Quadro.

**Comment by Alexander-LIT** [comment]
Very cool, thank you. Added it to my conky :smiley:

**Comment by leoratte** [comment]
Thank you for the sensor reports so far.
Could you provide additional reports with a fan, a temperature sensor and your high flow connected to the pump?

**Comment by Alexander-LIT** [comment]
> Thank you for the sensor reports so far. Could you provide additional reports with a fan, a temperature sensor and your high flow connected to the pump?

Certainly. I should be able to do that on the weekend. Possibly a little earlier if I find the time.

**Comment by Alexander-LIT** [comment]
Sure.

**Comment by Alexander-LIT** [comment]
> Great! Can you please also test connecting your flow sensor to the Flow 2 position on the Aquaero? I'd like to confirm its position in the report and implement it as well.

This just reminded me.... do the virtual sensors work as well? There is Virtual Sensor 1-8 in linux, which likely corresponds to "Software Sensor" in Aquasuite (not sure what they do, need to look it up).

Then there are 4 "Virtual temperature sensors" in Aquasuite which can be used for all kinds of calculations. I set one up to calculate the difference between Ambient and Case temperature (4.7° in the capture I uploaded). These are not currently displayed in my terminal sensor output.

Can these be pulled in as well? And if they can, I assume you would like me to create the max amount of sensors so can get some data from them?

**Comment by aleksamagicka** [comment]
Looks like I was wrong about that for the Aquaero. Virtual temp sensors are getting saved to the device; the first one is at `0x95` in the sensor report. I'll have to fix that here and rename "virtual" to "software" in mainline as well.

**Comment by Alexander-LIT** [comment]
I did some more captures of the Aquero.

Created the maximum of 4 virtual sensors. Sensor 2,3 and 4 equal sensor 1 + 10° offset on each so they should be hopefully easy to find.

I also did 2 captures of the two alarm states "warning" and "alert".

Will try get the pump values a bit later.

[aq.zip](https://github.com/aleksamagicka/aquacomputer_d5next-hwmon/files/10142225/aq.zip)

**Comment by Alexander-LIT** [comment]
Here we go with the data for the pump. Temp, flow and a fan connected to it. I also triggered an alert, but it does not show up in HWinfo.

[pump.zip](https://github.com/aleksamagicka/aquacomputer_d5next-hwmon/files/10143562/pump.zip)

**Comment by Alexander-LIT** [comment]
This is what I see in Data quick view, torque seems zero:
![2022-12-02_21-01](https://user-images.githubusercontent.com/63505691/205376086-a6479070-8e5d-404c-ace9-4174e4185144.png)

**Comment by aleksamagicka** [comment]
> I did some more captures of the Aquero.
> 
> Created the maximum of 4 virtual sensors. Sensor 2,3 and 4 equal sensor 1 + 10° offset on each so they should be hopefully easy to find.
> 
> I also did 2 captures of the two alarm states "warning" and "alert".
> 
> Will try get the pump values a bit later.
> 
> [aq.zip](https://github.com/aleksamagicka/aquacomputer_d5next-hwmon/files/10142225/aq.zip)

Virt temp sensors (Aquaero) are: 4.65, 14.65, 24.65, 34.65. `0x95` is the position of the first one - confirmed that, thanks! Will implement tomorrow hopefully.

> edit: Oh, and flow sensor was connected to flow 2 for the first capture as well

First flow sensor now reads 0, but the second one should be `1270`, which is exactly the same as on your picture. Offset of the first is `0xF9` and the second one is two bytes after: `0xFB`. Excellent, thank you! Will implement that as well.

I'll look at the other captures later on (I've got a few exams till the 10th so maybe it'll get delayed, but I'll do my best).

**Comment by Alexander-LIT** [comment]
Cool! Glad I could help!

**Comment by aleksamagicka** [comment]
Two flow sensors are supported in #55, please test when you find the time.

**Comment by Alexander-LIT** [comment]
I believe I checked  out the aquaer-two-flow-sensors branch correctly, but the flow sensor does not show up in "sensors" when it's connected to sensor 2.

**Comment by aleksamagicka** [comment]
My mistake, sorry. Didn't mark it as visible. Pushed a fix just now.

**Comment by Alexander-LIT** [comment]
Sucess! :D

aquaero-hid-3-8
Adapter: HID adapter
Fan 1 voltage:         12.10 V  
Fan 2 voltage:         12.28 V  
Fan 3 voltage:         12.19 V  
Fan 4 voltage:          0.00 V  
Fan 1 speed:           909 RPM  (min =  300 RPM, max = 1800 RPM)
Fan 2 speed:           398 RPM  (min =  300 RPM, max = 1800 RPM)
Fan 3 speed:           672 RPM  (min =  500 RPM, max = 1500 RPM)
Fan 4 speed:             0 RPM  (min =  450 RPM, max = 2000 RPM)
Flow sensor 1 [dL/h]:    0 RPM
Flow sensor 2 [dL/h]: 1255 RPM
Sensor 1:              +29.6°C  
Sensor 2:                  N/A  
Sensor 3:              +43.0°C  
Sensor 4:                  N/A  
Sensor 5:                  N/A  
Sensor 6:                  N/A  
Sensor 7:                  N/A  
Sensor 8:                  N/A  
Virtual sensor 1:      +50.0°C  
Virtual sensor 2:      +50.0°C  
Virtual sensor 3:      +50.0°C  
Virtual sensor 4:      +50.0°C  
Virtual sensor 5:      +50.0°C  
Virtual sensor 6:      +50.0°C  
Virtual sensor 7:      +50.0°C  
Virtual sensor 8:      +50.0°C  
Fan 1 power:            2.11 W  
Fan 2 power:            0.00 W  
Fan 3 power:            1.18 W  
Fan 4 power:            0.00 W  
Fan 1 current:        175.00 mA 
Fan 2 current:          0.00 A  
Fan 3 current:         97.00 mA 
Fan 4 current:          0.00 A

**Comment by aleksamagicka** [comment]
Merged, thanks.

**Comment by Alexander-LIT** [comment]
> Support for four calculated virtual temp sensors for the Aquaero is in #56. Please test, it seems to work fine on my Aquaero 6 LT.

Looks good to me!

aquaero-hid-3-8
Adapter: HID adapter
Fan 1 voltage:           12.07 V  
Fan 2 voltage:           12.25 V  
Fan 3 voltage:           12.16 V  
Fan 4 voltage:            0.00 V  
Fan 1 speed:            1021 RPM  (min =  300 RPM, max = 1800 RPM)
Fan 2 speed:             598 RPM  (min =  300 RPM, max = 1800 RPM)
Fan 3 speed:             762 RPM  (min =  500 RPM, max = 1500 RPM)
Fan 4 speed:               0 RPM  (min =  450 RPM, max = 2000 RPM)
Flow sensor 1 [dL/h]:   1260 RPM
Flow sensor 2 [dL/h]:      0 RPM
Sensor 1:                +31.1°C  
Sensor 2:                +25.8°C  
Sensor 3:                +43.6°C  
Sensor 4:                    N/A  
Sensor 5:                    N/A  
Sensor 6:                    N/A  
Sensor 7:                    N/A  
Sensor 8:                    N/A  
Virtual sensor 1:        +50.0°C  
Virtual sensor 2:        +50.0°C  
Virtual sensor 3:        +50.0°C  
Virtual sensor 4:        +50.0°C  
Virtual sensor 5:        +50.0°C  
Virtual sensor 6:        +50.0°C  
Virtual sensor 7:        +50.0°C  
Virtual sensor 8:        +50.0°C  
Calc. virtual sensor 1:   +5.3°C  
Calc. virtual sensor 2:  +15.3°C  
Calc. virtual sensor 3:  +25.3°C  
Calc. virtual sensor 4:  +35.3°C  
Fan 1 power:              2.59 W  
Fan 2 power:            640.00 mW 
Fan 3 power:              1.52 W  
Fan 4 power:              0.00 W  
Fan 1 current:          215.00 mA 
Fan 2 current:           53.00 mA 
Fan 3 current:          125.00 mA 
Fan 4 current:            0.00 A

**Comment by aleksamagicka** [comment]
Thanks! Merged.

**Comment by aleksamagicka** [comment]
I pushed some more commits to #52. I added the fan sensors @leoratte found and also the pump frequency (at least I think that's it). I also added the three flow sensors. Please test and report back which ones work and if something is possibly broken.

**Comment by Alexander-LIT** [comment]
I might be doing something wrong, but the Aquastream Ultimate branch is not showing my pump.

**Comment by aleksamagicka** [comment]
Sorry for that, I rebased it earlier using SmartGit and looks like it decided to remove the where the pump is marked as supported (I don't remember being prompted to do so). Force pushed it just now, please try again.

**Comment by Alexander-LIT** [comment]
The pump is showing up now. I'll try to get to testing it in the evening.

**Comment by Alexander-LIT** [comment]
Here is the output I am getting with fan, flow and temp sensor connected:

aquastreamultimate-hid-3-12
Adapter: HID adapter
Fan +12V:          0.00 V  
Pump +12V:        12.19 V  
Fan speed:          0 RPM
Pump speed:      4781 RPM
Pump frequency:  31205 RPM
Pressure [mbar]:  304 RPM
Flow speed 1:    1269 RPM
Flow speed 2:    32767 RPM
Flow speed 3:    1269 RPM
Coolant temp:     +31.0°C  
External temp:    +24.3°C  
Fan power:         0.00 W  
Pump power:        8.24 W  
Fan current:       0.00 A  

Fan seems to be running at 100%.

**Comment by aleksamagicka** [comment]
OK, so flow speed 2 and 3 are dummy values, removed those. Also fan sensors should now be fixed. Pump frequency I'm not sure about, what does aquasuite say (in the data quick view)?

**Comment by Alexander-LIT** [comment]
Frequency seems to be moving around 79.4 to79.7Hz.
Not entirely sure if the other flow sensors are dummies. One might be the unlockable software flow sensor and there is another port on the back of the pump that is labeled "RPM". I'll need to look up if a flow sensor connects to that as well.

**Comment by aleksamagicka** [comment]
I'm looking at this from the manual:

![image](https://user-images.githubusercontent.com/10512545/206830024-acea8c1b-95cc-4e73-a8d5-f034814ef58a.png)

RPM seems to be output only:

![image](https://user-images.githubusercontent.com/10512545/206830054-0528e4a2-dd1b-4187-92bf-d86c4d331a9c.png)

Here's what I find later:

![image](https://user-images.githubusercontent.com/10512545/206830121-9af9255c-1554-48d5-8ba3-52f5609b76e7.png)

That mentions the external flow sensor and the virtual one, which must be purchased and if it's anything like the on the D5 Next (it says it's calibrated to DP Ultra), it's mot likely not worth looking into because it seems very inaccurate.

I see hwinfo64 exposing the flow 3, but the value is the same. Looks like they weren't sure either.

**Comment by aleksamagicka** [comment]
I can't seem to find pump frequency in the sensor report, seeing that it's not in hwinfo64 maybe aquasuite is calculating it...

**Comment by aleksamagicka** [comment]
I can't stop you, but from all accounts that I've read the virtual flow sensor is... not great. Personally I'd say that it's just a gimmick and would not recommend buying that if it's only to try to help this driver.

Can you please test the PR again? Some minor issues were fixed/optimized and I've removed the pump frequency reporting. If all goes well, I'd merge it.

**Comment by Alexander-LIT** [comment]
The current version of the aquastream-ultimate-support branch causes a hard lockup on my system. Can't switch to a tty, mouse and keyboard are frozen and system needs to be turned off via powerbutton.

**Comment by leoratte** [comment]
I'll look into it.

**Comment by aleksamagicka** [comment]
Thanks @leoratte.

@Alexander-LIT, my apologies for that, should now be fixed.

**Comment by Alexander-LIT** [comment]
Much better :-)

aquastreamultimate-hid-3-13
Adapter: HID adapter
Fan voltage:        11.47 V  
Pump +12V:          12.22 V  
Fan speed:         2029 RPM
Pump speed:        4791 RPM
Pressure [mbar]:    306 RPM
Flow speed [dL/h]: 1343 RPM
Coolant temp:       +29.2°C  
External temp:      +23.1°C  
Fan power:           1.16 W  
Pump power:          8.17 W  
Fan current:       102.00 mA 
Pump current:      669.00 mA

**Comment by aleksamagicka** [comment]
Great, thank you! I'll just rename "Pump +12V" to "Pump voltage" and merge.

**Comment by aleksamagicka** [comment]
This is now merged. If any issues arise, please report.
